File: src/hforchestra/analysis/pe_extractor.py
"""
Complete PE Header Extractor for binary analysis.

This module provides comprehensive PE header extraction capabilities for
analyzing Windows executable files and detecting potential malware.
"""

import pefile
import os
import hashlib
import array
import math
import struct
import json
from pathlib import Path
from typing import Dict, Any, List, Optional, Tuple
from datetime import datetime
import sys


class CompletePEHeaderExtractor:
    """Extract ALL PE header fields comprehensively."""
    
    def __init__(self):
        self.suspicious_apis = {
            'VirtualAlloc', 'VirtualAllocEx', 'CreateRemoteThread', 'WriteProcessMemory',
            'ReadProcessMemory', 'OpenProcess', 'CreateProcess', 'ShellExecute',
            'WinExec', 'system', 'CreateFile', 'WriteFile', 'ReadFile',
            'RegCreateKey', 'RegSetValue', 'InternetOpen', 'HttpOpenRequest',
            'URLDownloadToFile', 'GetProcAddress', 'LoadLibrary', 'GetModuleHandle',
            'NtCreateThreadEx', 'NtAllocateVirtualMemory', 'NtWriteVirtualMemory'
        }
        
        # Machine types mapping
        self.machine_types = {
            0x0: 'UNKNOWN',
            0x1d3: 'AM33',
            0x8664: 'AMD64',
            0x1c0: 'ARM',
            0xaa64: 'ARM64',
            0x1c4: 'ARMNT',
            0xebc: 'EFI_BYTE_CODE',
            0x14c: 'I386',
            0x200: 'IA64',
            0x9041: 'M32R',
            0x266: 'MIPS16',
            0x366: 'MIPSFPU',
            0x466: 'MIPSFPU16',
            0x1f0: 'POWERPC',
            0x1f1: 'POWERPCFP',
            0x166: 'R4000',
            0x5032: 'RISCV32',
            0x5064: 'RISCV64',
            0x5128: 'RISCV128',
            0x1a2: 'SH3',
            0x1a3: 'SH3DSP',
            0x1a6: 'SH4',
            0x1a8: 'SH5',
            0x1c2: 'THUMB',
            0x169: 'WCEMIPSV2'
        }
        
        # Subsystem types
        self.subsystem_types = {
            0: 'UNKNOWN',
            1: 'NATIVE',
            2: 'WINDOWS_GUI',
            3: 'WINDOWS_CUI',
            5: 'OS2_CUI',
            7: 'POSIX_CUI',
            8: 'NATIVE_WINDOWS',
            9: 'WINDOWS_CE_GUI',
            10: 'EFI_APPLICATION',
            11: 'EFI_BOOT_SERVICE_DRIVER',
            12: 'EFI_RUNTIME_DRIVER',
            13: 'EFI_ROM',
            14: 'XBOX',
            16: 'WINDOWS_BOOT_APPLICATION'
        }
    
    def get_entropy(self, data: bytes) -> float:
        """Calculate Shannon entropy of data."""
        if len(data) == 0:
            return 0.0
        
        occurrences = array.array('L', [0] * 256)
        for byte in data:
            occurrences[byte] += 1
        
        entropy = 0
        data_len = len(data)
        for count in occurrences:
            if count > 0:
                p_x = float(count) / data_len
                entropy -= p_x * math.log(p_x, 2)
        
        return entropy
    
    def get_file_hash(self, file_path: str) -> Dict[str, str]:
        """Calculate multiple hash types."""
        try:
            with open(file_path, 'rb') as f:
                data = f.read()
            
            return {
                'md5': hashlib.md5(data).hexdigest(),
                'sha1': hashlib.sha1(data).hexdigest(),
                'sha256': hashlib.sha256(data).hexdigest()
            }
        except Exception as e:
            return {'error': str(e)}
    
    def extract_dos_header_complete(self, pe) -> Dict[str, Any]:
        """Extract complete DOS header information."""
        dos_header = pe.DOS_HEADER
        return {
            'e_magic': hex(dos_header.e_magic),
            'e_cblp': dos_header.e_cblp,
            'e_cp': dos_header.e_cp,
            'e_crlc': dos_header.e_crlc,
            'e_cparhdr': dos_header.e_cparhdr,
            'e_minalloc': dos_header.e_minalloc,
            'e_maxalloc': dos_header.e_maxalloc,
            'e_ss': dos_header.e_ss,
            'e_sp': dos_header.e_sp,
            'e_csum': dos_header.e_csum,
            'e_ip': dos_header.e_ip,
            'e_cs': dos_header.e_cs,
            'e_lfarlc': dos_header.e_lfarlc,
            'e_ovno': dos_header.e_ovno,
            'e_res': list(dos_header.e_res),
            'e_oemid': dos_header.e_oemid,
            'e_oeminfo': dos_header.e_oeminfo,
            'e_res2': list(dos_header.e_res2),
            'e_lfanew': dos_header.e_lfanew
        }
    
    def extract_file_header_complete(self, pe) -> Dict[str, Any]:
        """Extract complete file header information."""
        file_header = pe.FILE_HEADER
        return {
            'Machine': hex(file_header.Machine),
            'MachineType': self.machine_types.get(file_header.Machine, 'UNKNOWN'),
            'NumberOfSections': file_header.NumberOfSections,
            'TimeDateStamp': file_header.TimeDateStamp,
            'TimeDateStampFormatted': datetime.fromtimestamp(file_header.TimeDateStamp).strftime('%Y-%m-%d %H:%M:%S'),
            'PointerToSymbolTable': file_header.PointerToSymbolTable,
            'NumberOfSymbols': file_header.NumberOfSymbols,
            'SizeOfOptionalHeader': file_header.SizeOfOptionalHeader,
            'Characteristics': hex(file_header.Characteristics),
            'CharacteristicsFlags': self.get_characteristics_flags(file_header.Characteristics)
        }
    
    def extract_optional_header_complete(self, pe) -> Dict[str, Any]:
        """Extract complete optional header information."""
        optional_header = pe.OPTIONAL_HEADER
        return {
            'Magic': hex(optional_header.Magic),
            'MajorLinkerVersion': optional_header.MajorLinkerVersion,
            'MinorLinkerVersion': optional_header.MinorLinkerVersion,
            'SizeOfCode': optional_header.SizeOfCode,
            'SizeOfInitializedData': optional_header.SizeOfInitializedData,
            'SizeOfUninitializedData': optional_header.SizeOfUninitializedData,
            'AddressOfEntryPoint': hex(optional_header.AddressOfEntryPoint),
            'BaseOfCode': hex(optional_header.BaseOfCode),
            'ImageBase': hex(optional_header.ImageBase),
            'SectionAlignment': optional_header.SectionAlignment,
            'FileAlignment': optional_header.FileAlignment,
            'MajorOperatingSystemVersion': optional_header.MajorOperatingSystemVersion,
            'MinorOperatingSystemVersion': optional_header.MinorOperatingSystemVersion,
            'MajorImageVersion': optional_header.MajorImageVersion,
            'MinorImageVersion': optional_header.MinorImageVersion,
            'MajorSubsystemVersion': optional_header.MajorSubsystemVersion,
            'MinorSubsystemVersion': optional_header.MinorSubsystemVersion,
            'Win32VersionValue': optional_header.Win32VersionValue,
            'SizeOfImage': optional_header.SizeOfImage,
            'SizeOfHeaders': optional_header.SizeOfHeaders,
            'CheckSum': optional_header.CheckSum,
            'Subsystem': hex(optional_header.Subsystem),
            'SubsystemType': self.subsystem_types.get(optional_header.Subsystem, 'UNKNOWN'),
            'DllCharacteristics': hex(optional_header.DllCharacteristics),
            'DllCharacteristicsFlags': self.get_dll_characteristics_flags(optional_header.DllCharacteristics),
            'SizeOfStackReserve': optional_header.SizeOfStackReserve,
            'SizeOfStackCommit': optional_header.SizeOfStackCommit,
            'SizeOfHeapReserve': optional_header.SizeOfHeapReserve,
            'SizeOfHeapCommit': optional_header.SizeOfHeapCommit,
            'LoaderFlags': optional_header.LoaderFlags,
            'NumberOfRvaAndSizes': optional_header.NumberOfRvaAndSizes
        }
    
    def extract_data_directories_complete(self, pe) -> Dict[str, Any]:
        """Extract complete data directories information."""
        data_directories = pe.OPTIONAL_HEADER.DATA_DIRECTORY
        directories = {}
        
        for i, directory in enumerate(data_directories):
            if directory.Size > 0:
                directories[f'Directory_{i}'] = {
                    'VirtualAddress': hex(directory.VirtualAddress),
                    'Size': directory.Size
                }
        
        return directories
    
    def extract_sections_complete(self, pe) -> List[Dict[str, Any]]:
        """Extract complete section information."""
        sections = []
        for section in pe.sections:
            section_data = {
                'Name': section.Name.decode('utf-8').rstrip('\x00'),
                'VirtualAddress': hex(section.VirtualAddress),
                'VirtualSize': section.Misc_VirtualSize,
                'SizeOfRawData': section.SizeOfRawData,
                'PointerToRawData': section.PointerToRawData,
                'PointerToRelocations': section.PointerToRelocations,
                'PointerToLineNumbers': section.PointerToLineNumbers,
                'NumberOfRelocations': section.NumberOfRelocations,
                'NumberOfLineNumbers': section.NumberOfLineNumbers,
                'Characteristics': hex(section.Characteristics),
                'CharacteristicsFlags': self.get_section_characteristics_flags(section.Characteristics),
                'Entropy': self.get_entropy(section.get_data())
            }
            sections.append(section_data)
        
        return sections
    
    def extract_imports_complete(self, pe) -> Dict[str, Any]:
        """Extract complete import information."""
        imports = {}
        if hasattr(pe, 'DIRECTORY_ENTRY_IMPORT'):
            for entry in pe.DIRECTORY_ENTRY_IMPORT:
                dll_name = entry.dll.decode('utf-8')
                imports[dll_name] = []
                
                for imp in entry.imports:
                    import_info = {
                        'Name': imp.name.decode('utf-8') if imp.name else f'Ordinal_{imp.ordinal}',
                        'Ordinal': imp.ordinal,
                        'Address': hex(imp.address)
                    }
                    imports[dll_name].append(import_info)
        
        return imports
    
    def extract_exports_complete(self, pe) -> Dict[str, Any]:
        """Extract complete export information."""
        exports = {}
        if hasattr(pe, 'DIRECTORY_ENTRY_EXPORT'):
            for exp in pe.DIRECTORY_ENTRY_EXPORT.symbols:
                export_info = {
                    'Name': exp.name.decode('utf-8') if exp.name else f'Ordinal_{exp.ordinal}',
                    'Ordinal': exp.ordinal,
                    'Address': hex(exp.address)
                }
                exports[exp.name.decode('utf-8') if exp.name else f'Ordinal_{exp.ordinal}'] = export_info
        
        return exports
    
    def extract_resources_complete(self, pe) -> Dict[str, Any]:
        """Extract complete resource information."""
        resources = {}
        if hasattr(pe, 'DIRECTORY_ENTRY_RESOURCE'):
            for resource_type in pe.DIRECTORY_ENTRY_RESOURCE.entries:
                resource_type_name = resource_type.name if resource_type.name else str(resource_type.id)
                resources[resource_type_name] = []
                
                for resource_id in resource_type.directory.entries:
                    resource_id_name = resource_id.name if resource_id.name else str(resource_id.id)
                    resources[resource_type_name].append(resource_id_name)
        
        return resources
    
    def extract_complete_pe_headers(self, file_path: str) -> Dict[str, Any]:
        """Extract all PE header information comprehensively."""
        try:
            pe = pefile.PE(file_path)
            
            analysis = {
                'file_path': file_path,
                'file_size': os.path.getsize(file_path),
                'file_hashes': self.get_file_hash(file_path),
                'analysis_timestamp': datetime.now().isoformat(),
                'dos_header': self.extract_dos_header_complete(pe),
                'file_header': self.extract_file_header_complete(pe),
                'optional_header': self.extract_optional_header_complete(pe),
                'data_directories': self.extract_data_directories_complete(pe),
                'sections': self.extract_sections_complete(pe),
                'imports': self.extract_imports_complete(pe),
                'exports': self.extract_exports_complete(pe),
                'resources': self.extract_resources_complete(pe)
            }
            
            # Calculate summary statistics
            analysis['summary'] = self.calculate_summary_stats(analysis)
            
            pe.close()
            return analysis
            
        except Exception as e:
            return {
                'file_path': file_path,
                'error': str(e),
                'analysis_timestamp': datetime.now().isoformat()
            }
    
    def calculate_summary_stats(self, analysis: Dict[str, Any]) -> Dict[str, Any]:
        """Calculate summary statistics from the analysis."""
        summary = {
            'total_sections': len(analysis['sections']),
            'total_imports': sum(len(imports) for imports in analysis['imports'].values()),
            'total_exports': len(analysis['exports']),
            'suspicious_apis': [],
            'high_entropy_sections': []
        }
        
        # Check for suspicious APIs
        for dll, imports in analysis['imports'].items():
            for imp in imports:
                if imp['Name'] in self.suspicious_apis:
                    summary['suspicious_apis'].append({
                        'dll': dll,
                        'api': imp['Name']
                    })
        
        # Check for high entropy sections (potential packed/encrypted content)
        for section in analysis['sections']:
            if section['Entropy'] > 7.0:  # High entropy threshold
                summary['high_entropy_sections'].append({
                    'name': section['Name'],
                    'entropy': section['Entropy']
                })
        
        return summary
    
    def get_characteristics_flags(self, characteristics: int) -> List[str]:
        """Get human-readable characteristics flags."""
        flags = []
        if characteristics & 0x0001:
            flags.append('RELOCS_STRIPPED')
        if characteristics & 0x0002:
            flags.append('EXECUTABLE_IMAGE')
        if characteristics & 0x0004:
            flags.append('LINE_NUMS_STRIPPED')
        if characteristics & 0x0008:
            flags.append('LOCAL_SYMS_STRIPPED')
        if characteristics & 0x0010:
            flags.append('AGGRESSIVE_WS_TRIM')
        if characteristics & 0x0020:
            flags.append('LARGE_ADDRESS_AWARE')
        if characteristics & 0x0080:
            flags.append('BYTES_REVERSED_LO')
        if characteristics & 0x0100:
            flags.append('32BIT_MACHINE')
        if characteristics & 0x0200:
            flags.append('DEBUG_STRIPPED')
        if characteristics & 0x0400:
            flags.append('REMOVABLE_RUN_FROM_SWAP')
        if characteristics & 0x0800:
            flags.append('NET_RUN_FROM_SWAP')
        if characteristics & 0x1000:
            flags.append('SYSTEM')
        if characteristics & 0x2000:
            flags.append('DLL')
        if characteristics & 0x4000:
            flags.append('UP_SYSTEM_ONLY')
        if characteristics & 0x8000:
            flags.append('BYTES_REVERSED_HI')
        return flags
    
    def get_dll_characteristics_flags(self, dll_characteristics: int) -> List[str]:
        """Get human-readable DLL characteristics flags."""
        flags = []
        if dll_characteristics & 0x0020:
            flags.append('HIGH_ENTROPY_VA')
        if dll_characteristics & 0x0040:
            flags.append('DYNAMIC_BASE')
        if dll_characteristics & 0x0080:
            flags.append('FORCE_INTEGRITY')
        if dll_characteristics & 0x0100:
            flags.append('NX_COMPAT')
        if dll_characteristics & 0x0200:
            flags.append('NO_ISOLATION')
        if dll_characteristics & 0x0400:
            flags.append('NO_SEH')
        if dll_characteristics & 0x0800:
            flags.append('NO_BIND')
        if dll_characteristics & 0x1000:
            flags.append('APPCONTAINER')
        if dll_characteristics & 0x2000:
            flags.append('WDM_DRIVER')
        if dll_characteristics & 0x4000:
            flags.append('GUARD_CF')
        if dll_characteristics & 0x8000:
            flags.append('TERMINAL_SERVER_AWARE')
        return flags
    
    def get_section_characteristics_flags(self, characteristics: int) -> List[str]:
        """Get human-readable section characteristics flags."""
        flags = []
        if characteristics & 0x00000020:
            flags.append('CODE')
        if characteristics & 0x00000040:
            flags.append('INITIALIZED_DATA')
        if characteristics & 0x00000080:
            flags.append('UNINITIALIZED_DATA')
        if characteristics & 0x00000200:
            flags.append('INFO')
        if characteristics & 0x00000800:
            flags.append('REMOVE')
        if characteristics & 0x00001000:
            flags.append('COM_DAT')
        if characteristics & 0x00008000:
            flags.append('GPREL')
        if characteristics & 0x00020000:
            flags.append('MEM_PURGEABLE')
        if characteristics & 0x00020000:
            flags.append('MEM_16BIT')
        if characteristics & 0x00040000:
            flags.append('MEM_LOCKED')
        if characteristics & 0x00080000:
            flags.append('MEM_PRELOAD')
        if characteristics & 0x00100000:
            flags.append('ALIGN_1BYTES')
        if characteristics & 0x00200000:
            flags.append('ALIGN_2BYTES')
        if characteristics & 0x00300000:
            flags.append('ALIGN_4BYTES')
        if characteristics & 0x00400000:
            flags.append('ALIGN_8BYTES')
        if characteristics & 0x00500000:
            flags.append('ALIGN_16BYTES')
        if characteristics & 0x00600000:
            flags.append('ALIGN_32BYTES')
        if characteristics & 0x00700000:
            flags.append('ALIGN_64BYTES')
        if characteristics & 0x00800000:
            flags.append('ALIGN_128BYTES')
        if characteristics & 0x00900000:
            flags.append('ALIGN_256BYTES')
        if characteristics & 0x00A00000:
            flags.append('ALIGN_512BYTES')
        if characteristics & 0x00B00000:
            flags.append('ALIGN_1024BYTES')
        if characteristics & 0x00C00000:
            flags.append('ALIGN_2048BYTES')
        if characteristics & 0x00D00000:
            flags.append('ALIGN_4096BYTES')
        if characteristics & 0x00E00000:
            flags.append('ALIGN_8192BYTES')
        if characteristics & 0x01000000:
            flags.append('LNK_NRELOC_OVFL')
        if characteristics & 0x02000000:
            flags.append('MEM_DISCARDABLE')
        if characteristics & 0x04000000:
            flags.append('MEM_NOT_CACHED')
        if characteristics & 0x08000000:
            flags.append('MEM_NOT_PAGED')
        if characteristics & 0x10000000:
            flags.append('MEM_SHARED')
        if characteristics & 0x20000000:
            flags.append('MEM_EXECUTE')
        if characteristics & 0x40000000:
            flags.append('MEM_READ')
        if characteristics & 0x80000000:
            flags.append('MEM_WRITE')
        return flags
    
    def save_analysis_to_json(self, analysis: Dict[str, Any], output_path: str):
        """Save analysis results to JSON file."""
        with open(output_path, 'w') as f:
            json.dump(analysis, f, indent=2, default=str)
    
    def print_summary(self, analysis: Dict[str, Any]):
        """Print a summary of the analysis."""
        if 'error' in analysis:
            print(f"Error analyzing {analysis['file_path']}: {analysis['error']}")
            return
        
        print(f"\n=== PE Header Analysis Summary ===")
        print(f"File: {analysis['file_path']}")
        print(f"Size: {analysis['file_size']:,} bytes")
        print(f"MD5: {analysis['file_hashes']['md5']}")
        print(f"SHA256: {analysis['file_hashes']['sha256']}")
        
        print(f"\nMachine Type: {analysis['file_header']['MachineType']}")
        print(f"Subsystem: {analysis['optional_header']['SubsystemType']}")
        print(f"Sections: {analysis['summary']['total_sections']}")
        print(f"Imports: {analysis['summary']['total_imports']}")
        print(f"Exports: {analysis['summary']['total_exports']}")
        
        if analysis['summary']['suspicious_apis']:
            print(f"\nSuspicious APIs ({len(analysis['summary']['suspicious_apis'])}):")
            for api in analysis['summary']['suspicious_apis'][:5]:  # Show first 5
                print(f"  {api['dll']}:{api['api']}")
        
        if analysis['summary']['high_entropy_sections']:
            print(f"\nHigh Entropy Sections ({len(analysis['summary']['high_entropy_sections'])}):")
            for section in analysis['summary']['high_entropy_sections']:
                print(f"  {section['name']}: {section['entropy']:.2f}")
        
        print(f"\nAnalysis completed at: {analysis['analysis_timestamp']}") 

File: crates/analysis/src/pe_extractor.rs
//! Complete PE header extractor using goblin.

use anyhow::{Context, Result};
use chrono::Utc;
use goblin::pe::PE;
use md5::Md5;
use serde::{Deserialize, Serialize};
use sha1::Sha1;
use sha2::{Digest, Sha256};
use std::collections::{HashMap, HashSet};
use std::fs::File;
use std::io::Read;
use std::path::Path;

/// Section details in PE file.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SectionInfo {
    pub name: String,
    pub virtual_address: String,
    pub virtual_size: u32,
    pub raw_data_size: u32,
    pub entropy: f64,
    pub characteristics: String,
    pub flags: Vec<String>,
}

/// Summary statistics of the PE file.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct SummaryStats {
    pub total_sections: usize,
    pub total_imports: usize,
    pub total_exports: usize,
    pub suspicious_apis: Vec<SuspiciousApiCall>,
    pub high_entropy_sections: Vec<HighEntropySection>,
}

/// Suspicious API call descriptor.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SuspiciousApiCall {
    pub dll: String,
    pub api: String,
}

/// High entropy section descriptor.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HighEntropySection {
    pub name: String,
    pub entropy: f64,
}

/// Comprehensive analysis result of a PE binary.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PeAnalysis {
    pub file_path: String,
    pub file_size: u64,
    pub file_hashes: HashMap<String, String>,
    pub analysis_timestamp: String,
    pub machine: String,
    pub machine_type: String,
    pub number_of_sections: u16,
    pub entry_point: String,
    pub image_base: String,
    pub size_of_image: u32,
    pub subsystem: String,
    pub subsystem_type: String,
    pub sections: Vec<SectionInfo>,
    pub imports: HashMap<String, Vec<String>>,
    pub exports: Vec<String>,
    pub summary: SummaryStats,
}

/// Extracts detailed information from a Windows Portable Executable.
pub struct CompletePEHeaderExtractor {
    suspicious_apis: HashSet<String>,
    machine_types: HashMap<u16, String>,
    subsystem_types: HashMap<u16, String>,
}

impl Default for CompletePEHeaderExtractor {
    fn default() -> Self {
        Self::new()
    }
}

impl CompletePEHeaderExtractor {
    /// Create a new extractor.
    pub fn new() -> Self {
        let suspicious_apis = vec![
            "VirtualAlloc", "VirtualAllocEx", "CreateRemoteThread", "WriteProcessMemory",
            "ReadProcessMemory", "OpenProcess", "CreateProcess", "ShellExecute",
            "WinExec", "system", "CreateFile", "WriteFile", "ReadFile",
            "RegCreateKey", "RegSetValue", "InternetOpen", "HttpOpenRequest",
            "URLDownloadToFile", "GetProcAddress", "LoadLibrary", "GetModuleHandle",
            "NtCreateThreadEx", "NtAllocateVirtualMemory", "NtWriteVirtualMemory",
        ]
        .into_iter()
        .map(|s| s.to_string())
        .collect();

        let mut machine_types = HashMap::new();
        machine_types.insert(0x0, "UNKNOWN".to_string());
        machine_types.insert(0x8664, "AMD64".to_string());
        machine_types.insert(0x1c0, "ARM".to_string());
        machine_types.insert(0xaa64, "ARM64".to_string());
        machine_types.insert(0x14c, "I386".to_string());
        machine_types.insert(0x200, "IA64".to_string());

        let mut subsystem_types = HashMap::new();
        subsystem_types.insert(0, "UNKNOWN".to_string());
        subsystem_types.insert(1, "NATIVE".to_string());
        subsystem_types.insert(2, "WINDOWS_GUI".to_string());
        subsystem_types.insert(3, "WINDOWS_CUI".to_string());
        subsystem_types.insert(8, "NATIVE_WINDOWS".to_string());
        subsystem_types.insert(10, "EFI_APPLICATION".to_string());

        Self {
            suspicious_apis,
            machine_types,
            subsystem_types,
        }
    }

    /// Calculate entropy of a slice of bytes.
    pub fn get_entropy(&self, data: &[u8]) -> f64 {
        if data.is_empty() {
            return 0.0;
        }
        let mut occurrences = [0usize; 256];
        for &byte in data {
            occurrences[byte as usize] += 1;
        }
        let mut entropy = 0.0;
        let data_len = data.len() as f64;
        for &count in &occurrences {
            if count > 0 {
                let p_x = count as f64 / data_len;
                entropy -= p_x * p_x.log2();
            }
        }
        entropy
    }

    /// Calculate file hashes.
    pub fn get_file_hashes(&self, file_path: &Path) -> Result<HashMap<String, String>> {
        let mut file = File::open(file_path)?;
        let mut data = Vec::new();
        file.read_to_end(&mut data)?;

        let mut md5 = Md5::new();
        md5.update(&data);
        let md5_hash = hex::encode(md5.finalize());

        let mut sha1 = Sha1::new();
        sha1.update(&data);
        let sha1_hash = hex::encode(sha1.finalize());

        let mut sha256 = Sha256::new();
        sha256.update(&data);
        let sha256_hash = hex::encode(sha256.finalize());

        let mut hashes = HashMap::new();
        hashes.insert("md5".to_string(), md5_hash);
        hashes.insert("sha1".to_string(), sha1_hash);
        hashes.insert("sha256".to_string(), sha256_hash);

        Ok(hashes)
    }

    /// Extract all PE headers and information.
    pub fn extract_complete_pe_headers(&self, file_path: impl AsRef<Path>) -> Result<PeAnalysis> {
        let path = file_path.as_ref();
        let mut file = File::open(path)?;
        let mut buffer = Vec::new();
        file.read_to_end(&mut buffer)?;

        let pe = PE::parse(&buffer).context("Failed to parse PE binary")?;

        let file_hashes = self.get_file_hashes(path).unwrap_or_default();
        let file_size = path.metadata()?.len();

        let coff = pe.header.coff_header;
        let machine = format!("0x{:X}", coff.machine);
        let machine_type = self.machine_types.get(&coff.machine).cloned().unwrap_or_else(|| "UNKNOWN".to_string());
        let number_of_sections = coff.number_of_sections;

        let mut entry_point = "0x0".to_string();
        let mut image_base = "0x0".to_string();
        let mut size_of_image = 0u32;
        let mut subsystem = "0x0".to_string();
        let mut subsystem_type = "UNKNOWN".to_string();

        if let Some(opt) = pe.header.optional_header {
            entry_point = format!("0x{:X}", opt.standard_fields.address_of_entry_point);
            image_base = format!("0x{:X}", opt.windows_fields.image_base);
            size_of_image = opt.windows_fields.size_of_image;
            let subsystem_val = opt.windows_fields.subsystem;
            subsystem = format!("0x{:X}", subsystem_val);
            subsystem_type = self.subsystem_types.get(&subsystem_val).cloned().unwrap_or_else(|| "UNKNOWN".to_string());
        }

        // Section analysis
        let mut sections = Vec::new();
        for sec in &pe.sections {
            // Safe section name conversion
            let name = String::from_utf8_lossy(&sec.name)
                .trim_end_matches('\0')
                .to_string();

            // Extract section data from buffer to calculate entropy
            let start = sec.pointer_to_raw_data as usize;
            let end = (sec.pointer_to_raw_data + sec.size_of_raw_data) as usize;
            let sec_data = if start < buffer.len() && end <= buffer.len() {
                &buffer[start..end]
            } else {
                &[]
            };

            let entropy = self.get_entropy(sec_data);
            let flags = self.get_section_flags(sec.characteristics);

            sections.push(SectionInfo {
                name,
                virtual_address: format!("0x{:X}", sec.virtual_address),
                virtual_size: sec.virtual_size,
                raw_data_size: sec.size_of_raw_data,
                entropy,
                characteristics: format!("0x{:X}", sec.characteristics),
                flags,
            });
        }

        // Import analysis
        let mut imports = HashMap::new();
        let mut total_imports = 0;
        for imp in &pe.imports {
            let dll = imp.dll.to_string();
            let name = imp.name.to_string();
            imports.entry(dll).or_insert_with(Vec::new).push(name);
            total_imports += 1;
        }

        // Export analysis
        let mut exports = Vec::new();
        for exp in &pe.exports {
            if let Some(name) = exp.name {
                exports.push(name.to_string());
            }
        }

        // Calculate summary
        let mut suspicious_apis = Vec::new();
        for (dll, apis) in &imports {
            for api in apis {
                if self.suspicious_apis.contains(api) {
                    suspicious_apis.push(SuspiciousApiCall {
                        dll: dll.clone(),
                        api: api.clone(),
                    });
                }
            }
        }

        let mut high_entropy_sections = Vec::new();
        for sec in &sections {
            if sec.entropy > 7.0 {
                high_entropy_sections.push(HighEntropySection {
                    name: sec.name.clone(),
                    entropy: sec.entropy,
                });
            }
        }

        let summary = SummaryStats {
            total_sections: sections.len(),
            total_imports,
            total_exports: exports.len(),
            suspicious_apis,
            high_entropy_sections,
        };

        Ok(PeAnalysis {
            file_path: path.to_string_lossy().to_string(),
            file_size,
            file_hashes,
            analysis_timestamp: Utc::now().to_rfc3339(),
            machine,
            machine_type,
            number_of_sections,
            entry_point,
            image_base,
            size_of_image,
            subsystem,
            subsystem_type,
            sections,
            imports,
            exports,
            summary,
        })
    }

    /// Helper to convert section characteristic flags into readable strings.
    fn get_section_flags(&self, char: u32) -> Vec<String> {
        let mut flags = Vec::new();
        if char & 0x00000020 != 0 {
            flags.push("CNT_CODE".to_string());
        }
        if char & 0x00000040 != 0 {
            flags.push("CNT_INITIALIZED_DATA".to_string());
        }
        if char & 0x00000080 != 0 {
            flags.push("CNT_UNINITIALIZED_DATA".to_string());
        }
        if char & 0x20000000 != 0 {
            flags.push("MEM_EXECUTE".to_string());
        }
        if char & 0x40000000 != 0 {
            flags.push("MEM_READ".to_string());
        }
        if char & 0x80000000 != 0 {
            flags.push("MEM_WRITE".to_string());
        }
        flags
    }
}
